tasks/test_results_processor.py

```diff
@@ -107,167 +126,147 @@ def _bulk_write_tests_to_db(
-                        )
-                        + stmt.excluded.avg_duration_seconds
+                if outcome == str(Outcome.Pass):
+                    daily_totals[test_id]["pass_count"] += 1
+                elif outcome == str(Outcome.Failure) or outcome == str(Outcome.Error):
+                    daily_totals[test_id]["fail_count"] += 1
+                elif outcome == str(Outcome.Skip):
+                    daily_totals[test_id]["skip_count"] += 1
+
+            def create_daily_total():
+                daily_totals[test_id] = {
+                    "test_id": test_id,
+                    "repoid": repoid,
+                    "last_duration_seconds": duration_seconds,
+                    "avg_duration_seconds": duration_seconds,
+                    "pass_count": 1 if outcome == str(Outcome.Pass) else 0,
+                    "fail_count": 1
+                    if outcome == str(Outcome.Failure) or outcome == str(Outcome.Error)
+                    else 0,
+                    "skip_count": 1 if outcome == str(Outcome.Skip) else 0,
+                    "flaky_fail_count": 1
+                    if test_id in flaky_test_set and outcome == str(Outcome.Failure)
+                    else 0,
+                    "branch": branch,
+                    "date": date.today(),
+                    "latest_run": datetime.now(),
+                    "commits_where_fail": [commitid]
+                    if (
+                        outcome == str(Outcome.Failure) or outcome == str(Outcome.Error)
                     )
-                    / (
-                        DailyTestRollup.__table__.c.pass_count
-                        + DailyTestRollup.__table__.c.fail_count
-                        + 1
-                    ),
-                    "latest_run": stmt.excluded.latest_run,
-                    "pass_count": DailyTestRollup.__table__.c.pass_count
-                    + stmt.excluded.pass_count,
-                    "skip_count": DailyTestRollup.__table__.c.skip_count
-                    + stmt.excluded.skip_count,
-                    "fail_count": DailyTestRollup.__table__.c.fail_count
-                    + stmt.excluded.fail_count,
-                    "commits_where_fail": DailyTestRollup.__table__.c.commits_where_fail
-                    + stmt.excluded.commits_where_fail,
-                },
-            )
+                    else [],
+                }
+
+            if outcome != str(Outcome.Skip):
+                if test_id in daily_totals:
+                    update_daily_total()
+                else:
+                    create_daily_total()
+
+        # Save Tests
+        insert_on_conflict_do_nothing = (
+            insert(Test.__table__).values(test_data).on_conflict_do_nothing()
+        )
+        db_session.execute(insert_on_conflict_do_nothing)
+        db_session.flush()
+
+        # Upsert Daily Test Totals
+        rollup_table = DailyTestRollup.__table__
```

Comment:
I may need you to run me through this on Monday 😅 

Particularly why we need to do the calculations again on conflict; I figured they'd be computed in daily_totals

Reply:
ah the reason we do them is because the dictionary is just an intermediate holding container, if we came across two  test instances that map to the same rollup, we can't rely on the database to do the aggregating for us because there's a restriction that an `insert on conflict do update` can't have two entries that would insert/update the same row. So in the dictionary above we're duplicating this logic